File: javascript/tests/terbilang.test.ts
import { test } from 'uvu';
import * as assert from 'uvu/assert';
import { terbilang } from '../src/index';

test('should return correct words', () => {
  const result = terbilang(1_234_567);
  assert.equal(result, 'satu juta dua ratus tiga puluh empat ribu lima ratus enam puluh tujuh');
});

test('should parse string input', () => {
  const result = terbilang('1234567');
  assert.equal(result, 'satu juta dua ratus tiga puluh empat ribu lima ratus enam puluh tujuh');
});

test('minus', () => {
  assert.equal(terbilang(-1_234), 'minus seribu dua ratus tiga puluh empat');
});

test('nol', () => {
  assert.equal(terbilang(0), 'nol');
});

test('parse seratus with nol in long number', () => {
  assert.equal(terbilang(130_000_000), 'seratus tiga puluh juta');
});

test('parse seribu with nol in long number', () => {
  assert.equal(terbilang(1_000), 'seribu');
});

test('parse seratus', () => {
  assert.equal(terbilang(1_00), 'seratus');
});

test('parse triliun', () => {
  assert.equal(
    terbilang(91_234_567_912_345),
    'sembilan puluh satu triliun dua ratus tiga puluh empat miliar lima ratus enam puluh tujuh juta sembilan ratus dua belas ribu tiga ratus empat puluh lima',
  );
});

test('error', () => {
  // @ts-expect-error test purposes
  assert.throws(() => terbilang(true), 'argumen pertama harus merupakan angka ataupun string dan tidak boleh kosong');
});

test.run();

File: go/terbilang_test.go
package manusier_test

import (
	"testing"

	manusier "github.com/teknologi-umum/manusier/go"
)

func TestTerbilang_Normal(t *testing.T) {
	r := manusier.Terbilang(1234567)
	if r != "satu juta dua ratus tiga puluh empat ribu lima ratus enam puluh tujuh" {
		t.Error("should get satu juta dua ratus tiga puluh empat ribu lima ratus enam puluh tujuh. instead got:", r)
	}
}

func TestTerbilang_Minus(t *testing.T) {
	r := manusier.Terbilang(-1234)
	if r != "minus seribu dua ratus tiga puluh empat" {
		t.Error("should get minus seribu dua ratus tiga puluh empat. instead got:", r)
	}
}

func TestTerbilang_Nol(t *testing.T) {
	r := manusier.Terbilang(0)
	if r != "nol" {
		t.Error("should get nol. instead got:", r)
	}
}

func TestTerbilang_SeratusLong(t *testing.T) {
	r := manusier.Terbilang(130_000_000)
	if r != "seratus tiga puluh juta" {
		t.Error("should get seratus tiga puluh juta. instead got:", r)
	}
}

func TestTerbilang_Seratus(t *testing.T) {
	r := manusier.Terbilang(100)
	if r != "seratus" {
		t.Error("should get seratus. instead got:", r)
	}
}

func TestTerbilang_Seribu(t *testing.T) {
	r := manusier.Terbilang(1000)
	if r != "seribu" {
		t.Error("should get seribu. instead got:", r)
	}
}

func TestTerbilang_Triliun(t *testing.T) {
	r := manusier.Terbilang(91_234_567_912_345)
	if r != "sembilan puluh satu triliun dua ratus tiga puluh empat miliar lima ratus enam puluh tujuh juta sembilan ratus dua belas ribu tiga ratus empat puluh lima" {
		t.Error("should get sembilan puluh satu triliun dua ratus tiga puluh empat miliar lima ratus enam puluh tujuh juta sembilan ratus dua belas ribu tiga ratus empat puluh lima. instead got:", r)
	}
}
